[Mailer] added tag/metadata support

kbond opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #35047
| License       | MIT
| Doc PR        | todo

This is an alternative to #34766 for adding tag and metadata support in a more generalized way.

Most transports allow for open/click tracking headers - maybe this should be handled in a similar way?

I added implementations for the Postmark (SMTP and API) and Mailgun (SMTP and API) transports. I can add others and tests/docs if this is acceptable.

### Example:

```php
use Symfony\Component\Mailer\Header\MetadataHeader;
use Symfony\Component\Mailer\Header\TagHeader;

$email->getHeaders()->add(new TagHeader('password-reset'));
$email->getHeaders()->add(new MetadataHeader('Color', 'blue'));
$email->getHeaders()->add(new MetadataHeader('Client-ID', '12345'));
```

The Postmark/Mailgun providers will parse these into their own headers/payload. For transports that don't support tags/metadata, these are just added as custom headers:

```
X-Tag: password-reset
X-Metadata-Color: blue
X-Metadata-Client-ID: 12345
```

kbond committed:
[Mailer] added tag/metadata support

kbond commented:
Replying to https://github.com/symfony/symfony/issues/35047#issuecomment-567526041:

My experience is only with Postmark but I was able to find a similar concept for these 3rd party providers:

**Postmark**
Tag: single value
Metadata: key/value pair

**Mailgun**
Tag: single value
Metadata: key/value pair, referred to as "variables"

**SendGrid**
Tag: multiple values, referred to as "categories"
Metadata: key/value pair, referred to as "custom_args"

**Amazon**
Looks like tags/metadata is combined as "Tags"

AndreiIgna commented:
This is great and much better implementation than https://github.com/symfony/symfony/pull/34766 I was focusing on Postmark (using this atm) and that code targets only that transport.

PostmarkApiTransport doesn't have a `sendEmail` test, maybe you can add the one from my PR https://github.com/symfony/symfony/pull/34766/files#diff-e50347a040c466692646cd1e6b7fe257

kbond commented:
I have thought of some alternate solutions to account for the provider differences. Both would require knowledge of the transport when creating the email (which doesn't seem unreasonable).

1. A single `MetadataHeader` that takes a free form array. Providers would have knowledge of how to parse this array. This would be added as a json encoded `X-Metadata` header for providers that don't support this. This is the most generic solution but complex from a usage perspective.

    ```php
    $email->getHeaders()->addMetadata([
        'tag' => 'password-reset',
        'metadata' => [
            'Color' => 'blue',
            'Client-Id' => '12345',
        ],
    ]);
    ```

2. Each provider can define it's own header(s) (that would live in the bridge): `PostmarkHeader`, `MailgunHeader`, etc.

    ```php
    $email->getHeaders()->add((new PostmarkHeader())
        ->tag('password-reset')
        ->metadata([
            'Color' => 'blue',
            'Client-Id' => '12345',
        ])
    );
    ```

    Or multiple headers:

    ```php
    $email->getHeaders()->add(new PostmarkTag('password-reset'));
    $email->getHeaders()->add(new PostmarkMetadata([
        'Color' => 'blue',
        'Client-Id' => '12345',
    ]));
    ```

    This would allow us to utilize all the different features of each provider in a way that is easy for the user to use.

My vote would be for *option 2*.

AndreiIgna commented:
Hmm I think option 2 adds much more complexity.
I think option 1 or current implementation in PR would work better, because it doesn't require custom code for providers in app's code.

For example, with option 2 emails are created and sent by adding specific tags "PostmarkTag/PostmarkHeader". If email transport is changed later for a project, instead of changing the project's config now it requires code changes too.
At the moment I use Postmark for a project, but may change this to Sendgrid later. I'm using Mailer exactly for this reason, to change only the config once, and don't worry about how emails are created.

More generic `Headers` like Tag or Metadata are better in this case, and each transport uses this as it sees fit.

kbond commented:
It looks like SendGrid has more options than just tags and metadata, they have personalizations and substitutions. The current PR implementation would not be able to provide these options. Option 1 would but if you switched providers, it would likely silently break in a way that isn't obvious. Option 2 would also break but in a more obvious way (there would be an exception because header class would be missing).

Maybe the best course is to stick with the current PR implementation since multiple providers can support it and possibly add custom provider headers later?

fabpot commented on src/Symfony/Component/Mailer/Bridge/Mailgun/Transport/MailgunSmtpTransport.php:
```diff
@@ -27,4 +33,35 @@ public function __construct(string $username, string $password, string $region =
+            }
+        }
+
+        if (!empty($metadata)) {
```
`if ($metadata)`

fabpot commented on src/Symfony/Component/Mime/Header/Headers.php:
```diff
@@ -122,6 +122,22 @@ public function addParameterizedHeader(string $name, string $value, array $param
+    {
+        return $this->add(new MetadataHeader($key, $value));
+    }
+
```
I would add the header classes in the Mailer component and not add anything in the Mime component (as these headers are non-standard ones).

kbond commented:
I have moved the new Header classes to the Mailer component per @fabpot's request. I updated the PR description.

I have confirmed this works on Postmark but I don't use Mailgun. I'd appreciate someone with a Mailgun account to test.

If this is acceptable, I will add tests.

acasademont commented:
Sengrid user here, I'll take a look at this asap, it might fix #34700

fabpot commented on src/Symfony/Component/Mailer/Bridge/Mailgun/Transport/MailgunApiTransport.php:
```diff
@@ -115,6 +117,20 @@ private function getPayload(Email $email, Envelope $envelope): array
             }
 
             $payload['h:'.$name] = $header->getBodyAsString();
```
This should be removed right?

fabpot commented on src/Symfony/Component/Mailer/Bridge/Mailgun/Transport/MailgunApiTransport.php:
```diff
@@ -115,6 +117,20 @@ private function getPayload(Email $email, Envelope $envelope): array
+                continue;
+            }
+
+            $payload['h:'.$name] = $header->toString();
```
You should probably keep (copy/paste) line 119 instead.

fabpot commented on src/Symfony/Component/Mailer/Bridge/Mailgun/Transport/MailgunSmtpTransport.php:
```diff
@@ -27,4 +33,35 @@ public function __construct(string $username, string $password, string $region =
+                $headers->remove($name);
+            }
+
+            if ($header instanceof MetadataHeader) {
```
Could be an `elseif`

fabpot reviewed (changes requested):
LGTM, minor comments (and a bug)

kbond commented:
I made the changes and added tests. The tests are failing on travis (deps=low). Can I get some direction on how to fix?

Again, I don't feel confident in the Mailgun implementation as I can't do a live test but Postmark works as expected.

stof commented:
Mandrill (Mailchimp) also support these features (multiple values are supported for tags). Here are their doc:

- https://mandrillapp.com/api/docs/messages.JSON.html#method=send for the endpoint used by our ApiTransport
- https://mandrill.zendesk.com/hc/en-us/articles/205582117-How-to-Use-SMTP-Headers-to-Customize-Your-Messages#tag-your-messages for the SMTP doc about tags
- https://mandrill.zendesk.com/hc/en-us/articles/205582117-How-to-Use-SMTP-Headers-to-Customize-Your-Messages#use-custom-metadata for the SMTP doc about metadata

fabpot commented:
To fix the tests, you need to require at Mailer 5.1 for the bridges so that the new headers exist.

kbond commented:
Ok, travis has been fixed.

fabpot commented:
@kbond Can you try to do the same for Mandrill with the docs suggested by @stof? It not, happy to merge as is and let someone else take over for the other transport providers.

kbond commented:
Ok, I added Mandrill support for this feature. Again, I have not "live" tested this.

Mailgun and Mandrill both have an `HttpTransport` - this PR does not add support to these. Since these transports send the raw headers, it should work if using the proper header names for each service.

fabpot commented:
@kbond I think having support in HttpTransport classes as you describe would be good for consistency. Last thing to do before merge :)

kbond commented:
Ok, added support for Mandrill/Mailgun Http Transports. I added 2 traits to reduce duplication.

fabpot commented:
Thank you @kbond.
